openpype/pipeline/file_collection.py

```diff
@@ -0,0 +1,274 @@
+    ):
+        return file_path
+
+    expected_files = []
+    for frame in range(int(frame_start), (int(frame_end) + 1)):
+        frame_file_path = os.path.join(
+            dirpath, formattable_string.format(frame)
+        )
+        # normalize path
+        frame_file_path = os.path.normpath(frame_file_path)
+
+        # make sure file exists if ensure_exists is enabled
+        if only_existing and not os.path.exists(frame_file_path):
+            continue
+
+        # add to expected files
+        expected_files.append(
+            # add normalized path
+            frame_file_path
+        )
+
+    return expected_files
+
+
+def generate_expected_filepaths(
+    frame_start,
+    frame_end,
+    path,
+):
+    """Generate expected files from path
+
+    Args:
+        frame_start (int): Start frame of the sequence
+        frame_end (int): End frame of the sequence
+        path (str): Path to generate expected files from
+
+    Returns:
+        Any[list[str], str]: List of expected absolute paths to files
+            (frames of a sequence) or path if single file
+    """
+    return collect_filepaths_from_sequential_path(
+        frame_start,
+        frame_end,
+        path,
+        only_existing=False,
+    )
+
+
+def collect_basenames_from_sequential_path(
+    file_path,
+    frame_start=None,
+    frame_end=None,
+    only_existing=False,
+):
+    """Collect files from sequential path
+
+    Args:
+        file_path (str): Absolute path with any sequential pattern (##, %02d)
+        frame_start (Optional[int]): Start frame of the sequence
+        frame_end (Optional[int]): End frame of the sequence
+        only_existing (Optional[bool]): Ensure that files exists.
+
+    Returns:
+        list[str]: List of collected file names
+            (frames of a sequence) or path if single file
+    """
+
+    collected_abs_filepaths = collect_filepaths_from_sequential_path(
+        file_path,
+        frame_start,
+        frame_end,
+        only_existing,
+    )
+
+    if isinstance(collected_abs_filepaths, list):
+        return [os.path.basename(f_) for f_ in collected_abs_filepaths]
+
+    # return single file in list
+    return [os.path.basename(collected_abs_filepaths)]
+
+
+def convert_filename_to_formattable_string(filename):
+    """Convert filename to formattable string
+
+    Args:
+        filename (str): Filename to convert
+
+    Returns:
+        Any[str, None]: Formattable string or None if not possible
+                        to convert
+    """
+    new_filename = None
+
+    if "#" in filename:
+        # use regex to convert #### to {:0>4}
+        def replace(match):
+            return "{{:0>{}}}".format(len(match.group()))
+        new_filename = re.sub("#+", replace, filename)
+
+    elif "%" in filename:
+        # use regex to convert %04d to {:0>4}
+        def replace(match):
+            return "{{:0>{}}}".format(match.group()[1:])
+        new_filename = re.sub("%\\d+d", replace, filename)
```

Comment:
Are we sure that this catches all we want?
```python
import re

def replace(match):
    return "{{:0>{}}}".format(match.group()[1:])
    
    
for filename in [
    "hello.%04d",
    "hello.%4d",
    "hello.%d",
    "hello.%1d"
]:
    new_filename = re.sub("%\\d+d", replace, filename)
    print(filename, "  -->  ", new_filename)
```

```
hello.%04d   -->   hello.{:0>04d}
hello.%4d   -->   hello.{:0>4d}
hello.%d   -->   hello.%d
hello.%1d   -->   hello.{:0>1d}
```

Reply:
Great catch! But my head still bit hurt with some of your examples.. why whould you use either `%4d` or `%d` - have never seen it on any dccs.. Here is the fix.. 

```
import re

for filename in [
    "hello.%04d",
    "hello.%4d",
    "hello.%d",
    "hello.%1d"
]:
    number = 20
    padding = re.search("%(\\d)+d", filename)
    padding = padding.group(1) if padding else 1
    print("padding", padding)
    new_filename = re.sub(
        "%.*d",
        "{{:0>{}}}".format(padding),
        filename
    )
    print(filename, "  -->  ", new_filename)
    print(filename % number, "  -->  ", new_filename.format(number))
    ```